### PR TITLE
ci: add pre-commit and run prettier in it

### PR DESCRIPTION
**Pull request by wxharry**
## Brief Information

This pull request is in the type of:

- [ ] bug fixing
- [ ] new feature
- [x] ci
- [ ] others

Related issues:

* #364 

## Details
* add husky to package.json devDependencies
* add .husky/pre-commit

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/hypertrons/hypertrons-crx/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/hypertrons/hypertrons-crx)
- [x] I have run `yarn run prettier` before creating this PR

## Others
* If the changes are approved, remember to update checklist in PR template

### CODE REVIEW

**Comment by tyn1998** [comment]
Hey @wxharry, thank you for this CI improvement. 

Could you please add/link some docs that tell a developer how to use husky in our repo? You can directly modify the "[Quickstart](https://github.com/hypertrons/hypertrons-crx#quickstart)" section of the README if you are willing to.

**Comment by tyn1998** [comment]
> * If the changes are approved, remember to update checklist in PR template

Feel free to modify the PR template in this PR~

File: README.md [hunk 1]
```diff
@@ -88,10 +88,12 @@ Please read [CONTRIBUTING](./CONTRIBUTING.md) if you are new here or not familia
 2. cd hypertrons-crx
 
 3. yarn install
+   
+4. yarn run prepare
```
Comment (tyn1998) [inline README.md:92]: Hey @wxharry, maybe we can stress that this command will **only be run once** after cloning the repo.

File: README.zh-CN.md [hunk 2]
```diff
@@ -91,10 +91,12 @@ Hypertrons 浏览器插件项目旨在通过直接往 GitHub 页面中插入各
 2. cd hypertrons-crx
 
 3. yarn install
+   
+4. yarn run prepare
```
Comment (tyn1998) [inline README.zh-CN.md:95]: Here, too.

**Review by tyn1998 — APPROVED** [review]

**Comment by tyn1998** [comment]
Hey @wxharry, I think things not work as expected for now.

1. I broke 4 lines indent and committed the changes
image

2. `yarn run prettier` by husky caused the code to be formatted as before
image

That is not we want. What we expect is when somebody commits bad formatted code the commit will fail, right?

So we may use `yarn run prettier:check` instead `yarn run prettier` and fail the commit if the check is not passed, as well as prompt the committer with info like "please run `yarn run prettier` to format the code first".

What do you think so?

**Comment by tyn1998** [comment]
I also wonder if there is a way to automatically run `yarn run prepare` after `yarn install` is finished.

**Comment by wxharry** [comment]
> Hey @wxharry, I think things not work as expected for now.
> 
> So we may use `yarn run prettier:check` instead `yarn run prettier` and fail the commit if the check is not passed, as well as prompt the committer with info like "please run `yarn run prettier` to format the code first".
> 
> What do you think so?

Hey! IMO, Either failing the commit or letting prettier to change files are ok for our case. However, `yarn run prettier:check` may fail a commit if there is an unformatted unstaged file which could be annoying to our developers. 

I think we could use this [option](https://prettier.io/docs/en/precommit.html#option-2-pretty-quickhttpsgithubcomazzpretty-quick) given by prettier on pre-commit: using [pretty-quick](https://github.com/azz/pretty-quick). It provides flags `--staged` and `--check` and also a better human readable log information.

**Comment by tyn1998** [comment]
> However, `yarn run prettier:check` may fail a commit if there is an unformatted unstaged file which could be annoying to our developers.

Got it.



> I think we could use this [option](https://prettier.io/docs/en/precommit.html#option-2-pretty-quickhttpsgithubcomazzpretty-quick) given by prettier on pre-commit: using [pretty-quick](https://github.com/azz/pretty-quick). It provides flags `--staged` and `--check` and also a better human readable log information.

That is great!

**Comment by wxharry** [comment]
> I also wonder if there is a way to automatically run `yarn run prepare` after `yarn install` is finished.

`yarn run prepare` would automatically run after `yarn install`. See [package.json | Yarn #scripts](https://classic.yarnpkg.com/lang/en/docs/package-json/#toc-scripts)

**Review by tyn1998 — APPROVED** [review]
Great!

image

**Comment by tyn1998** [comment]
I will update CONTRIBUTING.

**Review by tyn1998 — APPROVED** [review]

**Comment by tyn1998** [comment]
> `yarn run prepare` would automatically run after `yarn install`. See [package.json | Yarn #scripts](https://classic.yarnpkg.com/lang/en/docs/package-json/#toc-scripts)

image

Testified it on a fresh cloned hypercrx repo!

Everything looks great to me, I will merge this PR.

Thank you, @wxharry.

**Pull request merged**